torch/csrc/jit/script/sugared_value.cpp

```diff
@@ -148,6 +148,21 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
                          << " cannot be used as a tuple";
 }
 
+static bool isRecursive(const TypePtr& classType, const TypePtr& attrType) {
+  if (attrType->isSubtypeOf(classType)) {
+    return true;
+  }
+
+  // Recursively check contained types. We need to do this because a user may do
+  // A -> B -> A.
+  for (const auto& type : attrType->containedTypes()) {
+    if (isRecursive(classType, type)) {
```

Comment:
The naive search here seems like it could get really inefficient with any complicated hierarchies

Reply:
BFS would be faster given how types are structured. But like, how complicated are we talking lol—this search is linear in the size of the type DAG, so unless someone is making 100000-nested compound types I think we're ok.

Reply:
I guess we're doing a second traversal if it's a tuple type, so if we have deep nested tuples we may need to add manually unrolling.

Reply:
I think this is exponential in the worst case...

Reply:
If you have `N` numbered nodes where each node is connected to all nodes with a lower number